crates/common/tedge_config_macros/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+                #[doku(as = "PathBuf")]
+                #[serde(alias = "certfile")]
+                cert_file: Utf8PathBuf,
+
+                /// Path to the client private key
+                #[doku(as = "PathBuf")]
+                #[serde(alias = "keyfile")]
+                key_file: Utf8PathBuf,
+            }
+        },
+
+        external: {
+            bind: {
+                /// The port mosquitto binds to for external use
+                #[tedge_config(example = "8883")]
+                port: u16,
+
+                /// The address mosquitto binds to for external use
+                #[tedge_config(example = "0.0.0.0")]
+                address: IpAddr,
+
+                /// Name of the network interface which mosquitto limits incoming connections on
+                #[tedge_config(example = "wlan0")]
+                interface: String,
+            },
+
+            /// Path to a file containing the PEM encoded CA certificates that are
+            /// trusted when checking incoming client certificates
+            #[tedge_config(example = "/etc/ssl/certs")]
+            #[doku(as = "PathBuf")]
+            #[serde(alias = "capath")]
+            ca_path: Utf8PathBuf,
+
+            /// Path to the certificate file which is used by the external MQTT listener
+            #[tedge_config(note = "This setting shall be used together with `mqtt.external.key_file` for external connections.")]
+            #[tedge_config(example = "/etc/tedge/device-certs/tedge-certificate.pem")]
+            #[doku(as = "PathBuf")]
+            #[serde(alias = "certfile")]
+            cert_file: Utf8PathBuf,
+
+            /// Path to the key file which is used by the external MQTT listener
+            #[tedge_config(note = "This setting shall be used together with `mqtt.external.cert_file` for external connections.")]
+            #[tedge_config(example = "/etc/tedge/device-certs/tedge-private-key.pem")]
+            #[doku(as = "PathBuf")]
+            #[serde(alias = "keyfile")]
+            key_file: Utf8PathBuf,
+        }
+    }
+}
+
+fn default_device_key(location: &TEdgeConfigLocation) -> Utf8PathBuf {
+    location
+        .tedge_config_root_path()
+        .join("device-certs")
+        .join("tedge-private-key.pem")
+}
+
+fn default_device_cert(location: &TEdgeConfigLocation) -> Utf8PathBuf {
+    location
+        .tedge_config_root_path()
+        .join("device-certs")
+        .join("tedge-certificate.pem")
+}
+
+fn default_mqtt_port() -> NonZeroU16 {
+    NonZeroU16::try_from(1883).unwrap()
+}
+
+#[test]
+fn root_cert_path_default() {
+    let dto = TEdgeConfigDto::default();
+    let reader = TEdgeConfigReader::from_dto(&dto, &TEdgeConfigLocation);
+    assert_eq!(reader.az.root_cert_path, "/etc/ssl/certs");
+}
+
+#[test]
+fn writable_keys_can_be_parsed_from_aliases() {
+    let _: WritableKey = "az.mapper.timestamp".parse().unwrap();
+    let _: WritableKey = "azure.mapper.timestamp".parse().unwrap();
+}
```

Comment:
How can this guesses that "azure" is an alias while "az" is not - *without* a `define_tedge_config!`?

What am I missing? Is the test name simply misleading?

Reply:
Sorry, this should define its own tedge config, but this currently uses the "global" example at the top of the file.

Reply:
No worries. Having this test using the global tedge config is okay. This is more that all the types and the `FromStr` implementation are defined by the macro.